### ISSUE TITLE
sam local start-api cannot access localstack

### ISSUE THREAD

**Issue by clotodex** [aws/aws-sam-cli#4861]
I am getting `The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.`

My command:
`sam local start-api --env-vars local_env.json -t ./cdk.out/ApiStack.template.json --region eu-central-1
 --docker-network "<localstack-network-id>" --force-image-build`
 
 When I make a request that runs in my lambda, it fails to even reach localstack (no interaction logged on their side).
 
 I ran the lambda code in its own docker container and my code works, even through docker so the issue must be with sam I guess.
 So I don't get what I am doing wrong.
 
 sam version: `1.76.0'
 LOCALSTACK_ENDPOINT=""
 
 ---
 
 The code I use to connect to the bucket:
 ```rust
     let config = aws_config::load_from_env().await;
    // local development
    let mut config = aws_sdk_s3::config::Builder::from(&config);
    if let Ok(endpoint) = env::var("LOCALSTACK_ENDPOINT") {
        let ep = s3::Endpoint::immutable(endpoint.parse::<http::Uri>().unwrap());
        let cred = s3::Credentials::new("access", "secret", None, None, "custom");
        // Region is not used but is required by the SDK's API
        let region = s3::Region::new("eu-central-1");
        config = config
            .endpoint_resolver(ep)
            .region(region)
            .credentials_provider(cred);
    }
    let client = s3::Client::from_conf(config.build());
```

local_env.json
```json
{
    "Parameters": {
		"BUCKET_RAW_NAME": "rawbucket",
		"LOCALSTACK_ENDPOINT": "http://127.17.0.1:4566/"
    }
}
```
(I also tried 127.0.0.1, localhost, localstack, ...)

**Comment by hawflau** [comment]
Hey @clotodex, thanks for raising the issue.

Did you try setting `LOCALSTACK_ENDPOINT` to match `"<localstack-network-id>"` you supplied in the `sam local start-api` comment? And start your local stack container with `--network "<localstack-network-id>"`?

**Comment by clotodex** [comment]
Yes I did, and it does not work. Also the explicit `sam local invoke` has the same problem.
Maybe something relevant: the lambda is running in the emulated `public.ecr.aws/lambda/provided:al2-rapid-arm64` runtime on an x86 machine. I don't know why this would be an issue, but I am close to believing anything at this point.

**Comment by hawflau** [comment]
okay, so this is also new to me. Can you try follow the steps below?
- create a docker user-defined bridge network
```docker network create --driver bridge my-network```
- run the localstack container. Note that we use both `--network` and `--name`
```docker run --rm -it -p 4566:4566 -p 4510-4559:4510-4559 --network my-network --name localstack localstack/localstack```
- inspect `my-network` by running `docker network inspect my-network`. You should see something like this:
```
[
    {
        "Name": "my-network",
        "Id": "f16dc9c8258f51dd1bffaa572f97e85d9de90235b509a7bc980b09693ce0c46e",
        "Created": "2023-03-17T00:52:10.050401725Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.20.0.0/16",
                    "Gateway": "172.20.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "d99fa5dc0215faf4a2eb3f196e5126fd37750cb691d7691438560ccb2caf503d": {
                "Name": "localstack",
                "EndpointID": "20bf89471e739a84b5f2d6b2a6f2453696bd37661b456e190b1eb2603b88e5f9",
                "MacAddress": "02:42:ac:14:00:02",
                "IPv4Address": "172.20.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```
- In your `local_envs.json` file, set `LOCALSTACK_ENDPOINT` to `http://localstack:4566`. Note that the hostname should match the container name.
- Run `sam local start-api --env-vars local_env.json -t ./cdk.out/ApiStack.template.json --region eu-central-1 --docker-network "my-network" --force-image-build`

In short, both the localstack container and the lambda container should be connected to the same user-defined bridge network. And use the name of the localstack container as the hostname for the AWS SDK to connect.

Let me know if that helps!

**Comment by clotodex** [comment]
The network inspect is as you described (running on 172.21.0.X)

I get the result again: `Not an ApiResult: Error { code: "PermanentRedirect", message: "The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.", request_id: "...", s3_extended_request_id: "..."`

A few notes on the setup:
- I am running on an AMD x86_64 machine
- The lambdas run in ARM arm64 (as rust lambdas)
- I allow the emulation with `docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`

Otherwise I followed your example fully and arrive at the same problem as before (and without SAM, when I execute the lambda code locally, everything works with localstack)

Do you have potential reasons why this might not work? versions, arm emulation specific problems?
